src/ducks/Signals/chart/VisualBacktestChart.js

```diff
@@ -0,0 +1,170 @@
+import React from 'react'
+import {
+  ResponsiveContainer,
+  ComposedChart,
+  XAxis,
+  YAxis,
+  Tooltip,
+  ReferenceDot,
+  ReferenceLine
+} from 'recharts'
+import cx from 'classnames'
+import { generateMetricsMarkup } from './../../SANCharts/utils'
+import { formatNumber } from './../../../utils/formatting'
+import { getDateFormats, getTimeFormats } from '../../../utils/dates'
+import chartStyles from './../../SANCharts/Chart.module.scss'
+import sharedStyles from './../../SANCharts/ChartPage.module.scss'
+import styles from './SignalPreview.module.scss'
+
+export const getDataKeys = (signal = {}) => {
+  if (signal.active_addresses) {
+    return { metric: 'dailyActiveAddresses', signal: 'active_addresses' }
+  }
+  if (signal.price) return { metric: 'priceUsd', signal: 'price' }
+  return { metric: 'priceUsd' }
+}
+
+const mapWithTimeseries = items =>
+  items.map(item => ({ ...item, datetime: +new Date(item.datetime) }))
+
+const mapWithMidnightTime = date => {
+  const datetime = new Date(date)
+  datetime.setUTCHours(0, 0, 0, 0)
+  return +new Date(datetime)
+}
+
+const VisualBacktestChart = ({
+  triggeredSignals,
+  timeseries = [],
+  metrics,
+  label
+}) => {
+  const data = mapWithTimeseries(timeseries)
+  const dataKeys = getDataKeys(triggeredSignals[0])
+  const markup = generateMetricsMarkup(metrics)
+
+  const renderChart = () => {
+    return (
+      <ComposedChart data={data}>
+        <XAxis
+          dataKey='datetime'
+          type='number'
+          scale='time'
+          tick={false}
+          allowDataOverflow
+          domain={['dataMin', 'dataMax']}
+          hide
+        />
+
+        <YAxis hide domain={['auto', 'dataMax']} dataKey={dataKeys.metric} />
+
+        {markup}
+
+        {triggeredSignals.map((point, i) => {
+          let price
+          const isPrice = dataKeys.signal === 'price'
+          const date = isPrice
+            ? mapWithMidnightTime(point.datetime)
+            : +new Date(point.datetime)
+
+          if (isPrice) {
+            const item = data.find(({ datetime }) => datetime === date)
+            price = item ? item.priceUsd : point.price
+          }
+
+          return dataKeys.signal ? (
+            <ReferenceDot
+              x={date}
+              y={price || point[dataKeys.signal]}
```

Comment:
If you make the function, which will return y, you can combine 3 places from the code to one and avoid "let"